File: app/src/main/java/net/mbonnin/arcanetracker/QuitDetector.java
package net.mbonnin.arcanetracker;

import android.os.Handler;
import android.view.View;

import timber.log.Timber;

/**
 * Created by martin on 11/2/16.
 */

public class QuitDetector {
    private final MainViewCompanion mainViewCompanion;
    private final ViewManager viewManager;
    private final Settings settings;
    private Handler mHandler;
    private int[] mPosition = new int[2];

    private static int STATE_NOTHING = 0;
    private static int STATE_WAS_OPEN = 1;
    private static int STATE_WAS_CLOSED = 2;

    private int state = STATE_NOTHING;

    private final Runnable mCheckFullScreenRunnable = new Runnable() {

        @Override
        public void run() {

            View view = mainViewCompanion.getMainView();
            view.getLocationOnScreen(mPosition);

            //Timber.i("view %d - %d", mPosition[1], view.getHeight());
            if (mPosition[1] != 0) {
                if (settings.get(Settings.AUTO_QUIT, true)) {
                    if (state == STATE_NOTHING) {
                        Timber.i("Exit FullScreen detected");
                        state = mainViewCompanion.isOpen() ? STATE_WAS_OPEN: STATE_WAS_CLOSED;
                        mainViewCompanion.setOpen(false);
                        viewManager.removeAllViewsExcept(view);
                    }
                }
            } else {
                if (state != STATE_NOTHING) {
                    Timber.i("Back to FullScreen");
                    mainViewCompanion.setOpen(state == STATE_WAS_OPEN);
                    mainViewCompanion.show(true);
                    state = STATE_NOTHING;
                }
            }
            mHandler.postDelayed(this, 1000);
        }
    };

    public void start() {
        mHandler.postDelayed(mCheckFullScreenRunnable, 4000);
    }

    public void stop() {
        mHandler.removeCallbacks(mCheckFullScreenRunnable);
    }

    public QuitDetector(MainViewCompanion mainViewCompanion, ViewManager viewManager, Settings settings) {
        this.mainViewCompanion = mainViewCompanion;
        this.viewManager = viewManager;
        this.settings = settings;
        this.mHandler = new Handler();
    }
}

File: app/src/main/java/net/mbonnin/arcanetracker/QuitDetector.kt
package net.mbonnin.arcanetracker

import android.content.Context.WINDOW_SERVICE
import android.os.Build
import android.os.Handler
import android.provider.Settings.canDrawOverlays
import android.util.DisplayMetrics
import android.view.WindowManager
import net.mbonnin.arcanetracker.ui.overlay.Overlay
import net.mbonnin.arcanetracker.ui.settings.SettingsCompanion
import timber.log.Timber


class QuitDetector {
    private val mHandler = Handler()
    private var visible = true

    private val mCheckFullScreenRunnable = object : Runnable {

        override fun run() {
            val displayMetrics = DisplayMetrics()
            val windowManager = HDTApplication.get().getSystemService(WINDOW_SERVICE) as WindowManager
            windowManager.defaultDisplay.getMetrics(displayMetrics)
            val height = displayMetrics.heightPixels
            val width = displayMetrics.widthPixels

            if (height > width && visible) {
                Overlay.get().hide()
                visible = false

                Timber.d("Hide")

            } else if (height <= width && !visible) {
                Timber.d("Show")

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    if (!canDrawOverlays(HDTApplication.get())) {
                        // we might come here from the quit detector. In that case, just do nothing
                        return
                    }
                }

                Overlay.get().show()
                visible = true
            }

            if (!visible) {
                val timeoutMin = SettingsCompanion.getTimeoutValue()

                if (timeoutMin > 0 && System.currentTimeMillis() - lastTimeMillis > timeoutMin * 60 * 1000) {
                    Timber.d("Quit after inactivity")
                    Utils.exitApp()
                }
            }
            mHandler.postDelayed(this, 1000)
        }
    }

    fun start() {
        ping()
        mHandler.postDelayed(mCheckFullScreenRunnable, 4000)
    }

    @Volatile private var lastTimeMillis: Long = 0L

    fun ping() {
        lastTimeMillis = System.currentTimeMillis()
    }

    companion object {
        internal var sQuitDetector: QuitDetector? = null

        fun get(): QuitDetector {
            if (sQuitDetector == null) {
                sQuitDetector = QuitDetector()
            }
            return sQuitDetector!!
        }
    }
}
